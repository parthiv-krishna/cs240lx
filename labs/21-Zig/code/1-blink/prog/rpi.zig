extern fn nop() void;

// countdown 'ticks' cycles
pub fn delay(ticks: u32) void {
    var cnt = @as(u32, ticks);
    while (cnt > 0) {
        nop();
        cnt -= @as(u32, 1);
    }
}


