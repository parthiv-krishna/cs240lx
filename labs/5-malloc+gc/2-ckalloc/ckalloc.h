// weak test of malloc/free: checks that subsequent allocations are
// <sizeof (void *)> bytes from each other.

#ifndef __CKALLOC_H__
#define __CKALLOC_H__

#include "src-loc.h"

typedef enum {  ALLOCED = 11, FREED } state_t;

// pull the remainder into the second redzone.
typedef struct ck_hdr {
    struct ck_hdr *next;
    uint32_t nbytes_alloc;  // how much the user requested to allocate.
    uint32_t state;          // state of the block: { ALLOCED, FREED }

    src_loc_t alloc_loc;    // location they called ckalloc() at.

    // used for gc: i didn't cksum these.
    uint32_t refs_start;    // number of pointers to the start of the block.
    uint32_t refs_middle;   // number of pointers to the middle of the block.

    uint16_t mark;          // 0 initialize.
} hdr_t;

// returns pointer to the first header block.
hdr_t *ck_first_hdr(void);

// returns pointer to next hdr or 0 if none.
hdr_t *ck_next_hdr(hdr_t *p);

// get the number of bytes for hdr <h>
static inline unsigned ck_nbytes(hdr_t *h) {
    return h->nbytes_alloc;
}
// get the start of allocated data for <h>
void *ck_hdr_start(hdr_t *h);
// get the end of allocated data for <h>
void *ck_hdr_end(hdr_t *h);

// is <ptr> in block <h>?
unsigned 
ck_ptr_in_block(hdr_t *h, void *ptr);

// is pointer <p> on the allocated list?
int ck_ptr_is_alloced(void *ptr);

#define ckalloc(_n) (ckalloc)(_n, SRC_LOC_MK())
#define ckfree(_ptr) (ckfree)(_ptr, SRC_LOC_MK())
void *(ckalloc)(uint32_t nbytes, src_loc_t loc);
void (ckfree)(void *addr, src_loc_t loc);

#endif
