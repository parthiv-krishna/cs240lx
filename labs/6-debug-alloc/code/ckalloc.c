// implement a simple ckalloc/free that adds ckhdr_t to the 
// allocation.

#include "test-interface.h"
#include "ckalloc.h"

// keep a list of allocated blocks.
static hdr_t *alloc_list;

// returns pointer to the first header block.
hdr_t *ck_first_hdr(void) {
    return alloc_list;
}

// returns pointer to next hdr or 0 if none.
hdr_t *ck_next_hdr(hdr_t *p) {
    if(p)
        return p->next;
    return 0;
}


void *ck_hdr_start(hdr_t *h) {
    return &h[1];
}

// one past the last byte of allocated memory.
void *ck_hdr_end(hdr_t *h) {
    return (char *)ck_hdr_start(h) + ck_nbytes(h);
}

// is ptr in <h>?
unsigned ck_ptr_in_block(hdr_t *h, void *ptr) {
    char *start = (char *)(h + 1);
    char *end = start + h->nbytes_alloc;
    char *p = (char *)ptr;
    return (p >= start && p < end);
}


hdr_t *ck_ptr_is_alloced(void *ptr) {
    for(hdr_t *h = ck_first_hdr(); h; h = ck_next_hdr(h))
        if(ck_ptr_in_block(h,ptr)) {
            // output("found %p in %p\n", ptr, h);
            return h;
        }
    return NULL;
}

static void list_remove(hdr_t **l, hdr_t *h) {
    assert(l);
    hdr_t *prev = *l;
 
    if(prev == h) {
        *l = h->next;
        return;
    }

    hdr_t *p;
    while((p = ck_next_hdr(prev))) {
        if(p == h) {
            prev->next = p->next;
            return;
        }
        prev = p;
    }
    panic("did not find %p in list\n", h);
}

// free a block allocated with <ckalloc>
void (ckfree)(void *addr, src_loc_t l) {
    hdr_t *h = (void *)addr;
    h -= 1;

    if(h->state != ALLOCED)
        loc_panic(l, "freeing unallocated memory: state=%d\n", h->state);
    loc_debug(l, "freeing %p\n", addr);
    
    h->state = FREED;
    memset(h + 1, 0, h->nbytes_alloc);
    assert(ck_ptr_is_alloced(addr));
    kr_free(h);
    list_remove(&alloc_list, h);
    assert(!ck_ptr_is_alloced(addr));
}

// interpose on kr_malloc allocations and
//  1. allocate enough space for a header and fill it in.
//  2. add the allocated block to  the allocated list.
void *(ckalloc)(uint32_t nbytes, src_loc_t loc) {

    hdr_t *h = kr_malloc(nbytes + sizeof *h);

    memset(h, 0, sizeof *h);
    h->nbytes_alloc = nbytes;
    h->state = ALLOCED;
    h->alloc_loc = loc;


    loc_debug(loc, "allocating %p\n", h);

    void *addr = (void *)(h + 1);

    assert(!ck_ptr_is_alloced(addr));
    h->next = alloc_list;
    alloc_list = h;

    assert(ck_ptr_is_alloced(addr));

    return addr;
}
