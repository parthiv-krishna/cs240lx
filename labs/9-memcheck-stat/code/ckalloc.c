// implement a simple ckalloc/free that adds ckhdr_t to the 
// allocation.

#include "test-interface.h"
#include "ckalloc.h"

// keep a list of allocated blocks.
static hdr_t *alloc_list;
static hdr_t *free_list;

unsigned ck_verbose_p = 0;

// returns pointer to the first header block.
hdr_t *ck_first_hdr(void) {
    return alloc_list;
}

// returns pointer to next hdr or 0 if none.
hdr_t *ck_next_hdr(hdr_t *p) {
    if(p)
        return p->next;
    return 0;
}


void *ck_hdr_start(hdr_t *h) {
    return &h[1];
}

// one past the last byte of allocated memory.
void *ck_hdr_end(hdr_t *h) {
    return (char *)ck_hdr_start(h) + ck_nbytes(h);
}

// is ptr in <h>?
unsigned ck_ptr_in_block(hdr_t *h, void *ptr) {
    char *start = (char *)(h + 1);
    char *end = start + h->nbytes_alloc;
    char *p = (char *)ptr;
    return (p >= start && p < end);
}


hdr_t *ck_ptr_is_alloced(void *ptr) {
    for(hdr_t *h = ck_first_hdr(); h; h = ck_next_hdr(h))
        if(ck_ptr_in_block(h,ptr)) {
            // output("found %p in %p\n", ptr, h);
            return h;
        }
    return NULL;
}

static void list_remove(hdr_t **l, hdr_t *h) {
    assert(l);
    hdr_t *prev = *l;
 
    if(prev == h) {
        *l = h->next;
        return;
    }

    hdr_t *p;
    while((p = ck_next_hdr(prev))) {
        if(p == h) {
            prev->next = p->next;
            return;
        }
        prev = p;
    }
    panic("did not find %p in list\n", h);
}


void make_redzone(char *p, unsigned n) {
    memset(p, REDZONE_VAL, n);
}

int check_redzone(char *p, unsigned n) {
    for(unsigned i = 0; i < n; i++) {
        if(p[i] != REDZONE_VAL) {
            printk("found error at %p, %u, %d\n", p, i, p[i]);
            return 4*i + 1;
        }
    }
    return 0;
}

static int ckalloc_count = 1;

// free a block allocated with <ckalloc>
void (ckfree)(void *addr, src_loc_t l) {
    hdr_t *h = (void *)addr;
    h -= 1;

    int ofst = check_redzone(h->rz1, REDZONE_NBYTES);
    if (ofst != 0) {
        trace("ERROR:block %u  corrupted at offset %d\n",
              h->block_id, ofst - 1);
        trace("        logical block id=%u,  nbytes=%u\n", h->block_id, h->nbytes_alloc);
        trace("        Block allocated at: %s:%s:%d\n", h->alloc_loc.file, h->alloc_loc.func, h->alloc_loc.lineno);
        
        panic("corrupted block %u [addr=%p]", h->block_id, addr);
    }        
    
    ofst = check_redzone((char *)addr + h->nbytes_alloc, REDZONE_NBYTES);
    if (ofst != 0) {
        trace("ERROR:block %u  corrupted at offset %d\n",
              h->block_id, ofst - 1 + h->nbytes_alloc);
        trace("        logical block id=%u,  nbytes=%u\n", h->block_id, h->nbytes_alloc);
        trace("        Block allocated at: %s:%s:%d\n", h->alloc_loc.file, h->alloc_loc.func, h->alloc_loc.lineno);
        
        panic("corrupted block %u [addr=%p]", h->block_id, addr);
    }        

    

    if(h->state != ALLOCED)
        loc_panic(l, "freeing unallocated memory: state=%d\n", h->state);
    loc_debug(l, "freeing block=%u [addr=%p]\n", h->block_id, addr);
    
    h->state = FREED;
    make_redzone((char *)(h + 1), h->nbytes_alloc); // nuke data
    assert(ck_ptr_is_alloced(addr));

    list_remove(&alloc_list, h);
    h->next = free_list;
    free_list = h;


    // list_remove(&alloc_list, h);
    assert(!ck_ptr_is_alloced(addr));
}


// interpose on kr_malloc allocations and
//  1. allocate enough space for a header and fill it in.
//  2. add the allocated block to  the allocated list.
void *(ckalloc)(uint32_t nbytes, src_loc_t loc) {


    hdr_t *h = kr_malloc(nbytes + sizeof *h + REDZONE_NBYTES);

    memset(h, 0, sizeof *h);
    h->nbytes_alloc = nbytes;
    h->state = ALLOCED;
    h->alloc_loc = loc;
    h->block_id = ckalloc_count++;

    loc_debug(loc, "allocating block=%u [addr=%p]\n", h->block_id, h);

    void *addr = (void *)(h + 1);

    make_redzone(h->rz1, REDZONE_NBYTES);
    make_redzone((char *)addr + nbytes, REDZONE_NBYTES);

    assert(!ck_ptr_is_alloced(addr));
    h->next = alloc_list;
    alloc_list = h;

    assert(ck_ptr_is_alloced(addr));

    return addr;
}


int ck_heap_errors() {
    int errors = 0;
    int nblocks = 0;
    for(hdr_t *h = alloc_list; h; h = h->next) {
        nblocks++;

        if (h->state == ALLOCED) {
            if (ck_verbose_p) printk("alloc block block=%u [addr=[%p]\n", h->block_id, h);
            if (check_redzone(h->rz1, REDZONE_NBYTES) ||
                check_redzone((char *)(h + 1) + h->nbytes_alloc, REDZONE_NBYTES)) {
                    errors++;
            }
        } else {
            panic("free block %u [addr=%p] in alloc list", h->block_id, h);
        }
    }

    for(hdr_t *h = free_list; h; h = h->next) {
        nblocks++;

        if (h->state == FREED) {
            if (ck_verbose_p) printk("free block block=%u [addr=%p]\n", h->block_id, h);
            if (check_redzone(h->rz1, REDZONE_NBYTES) ||
                check_redzone((char *)(h + 1), h->nbytes_alloc) ||
                check_redzone((char *)(h + 1) + h->nbytes_alloc, REDZONE_NBYTES)) {
                    errors++;
            }
        } else {
            panic("alloc block %u [addr=%p] in free list", h->block_id, h);
        }

    }

    return errors;
}
