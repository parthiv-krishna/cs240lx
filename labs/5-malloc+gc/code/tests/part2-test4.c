// using a doubly linked list to make sure we can handle cycles.
#include "rpi.h"
#include "ckalloc.h"

#define N 100

// allocate a linked list.  do a leak detection.
void *test(void) {
    struct list {
        int x; 
        struct list *next;
        struct list *prev;
    } *h = 0;

    for(int i = 0; i < 100; i++) {
        struct list *e = ckalloc(sizeof *h);
        memset(e, 0, sizeof *e);
        e->x = 0;
        if(h) {
            assert(!h->prev);
            h->prev = e;
            e->next = h;
        }
        h = e;
    }
    // compiler optimizes this stuff away.
    // h= 0;
    // even this:
    //  memset(&h, 0, sizeof h);
    // perhaps if we put it in another file?
    check_no_leak();
    // return to try to get around compiler tricks.
    return h;
}

void notmain(void) {
    printk("GC test: checking that a doubly-linked list is handled.\n");
    // note: if we do a check after this, may miss b/c the value is laying around
    // on the stack or reg
    printk("last alloc =%p\n", test());
}
