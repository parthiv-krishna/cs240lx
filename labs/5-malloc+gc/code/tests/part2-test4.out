find-ttyusb.c:find_ttyusb_first:106:FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
my-install: tty-usb=</dev/ttyUSB0> program=<./part2-test4.bin> about to boot
******************sending 9760 bytes
waiting for a start
got a start
putting code
put code and got back
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
GC test: checking that a doubly-linked list is handled.
sizeof hdr=56, redzone=128
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1000b8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100200
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100348
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100490
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1005d8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100720
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100868
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1009b0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100af8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100c40
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100d88
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100ed0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x101018
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x101160
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1012a8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1013f0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x101538
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x101680
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1017c8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x101910
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x101a58
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x101ba0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x101ce8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x101e30
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x101f78
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1020c0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x102208
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x102350
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x102498
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1025e0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x102728
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x102870
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1029b8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x102b00
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x102c48
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x102d90
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x102ed8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x103020
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x103168
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1032b0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1033f8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x103540
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x103688
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1037d0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x103918
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x103a60
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x103ba8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x103cf0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x103e38
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x103f80
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1040c8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x104210
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x104358
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1044a0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1045e8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x104730
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x104878
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1049c0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x104b08
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x104c50
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x104d98
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x104ee0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x105028
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x105170
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1052b8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x105400
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x105548
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x105690
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1057d8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x105920
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x105a68
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x105bb0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x105cf8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x105e40
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x105f88
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1060d0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x106218
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x106360
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1064a8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1065f0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x106738
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x106880
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1069c8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x106b10
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x106c58
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x106da0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x106ee8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x107030
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x107178
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1072c0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x107408
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x107550
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x107698
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1077e0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x107928
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x107a70
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x107bb8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x107d00
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x107e48
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x107f90
TRACE:going to check heap: 32800 bytes allocated, 1015776 bytes left
TRACE:SUCCESS: checked 100 blocks, detected no errors
TRACE:GC: SUCCESS heap checked out
ck-gc.c:mark_all:211:stack has 34 words
---------------------------------------------------------
checking for leaks:
TRACE:	GC:Checked 100 blocks.
TRACE:		GC:SUCCESS: No leaks found!
----------------------------------------------------------
TRACE:GC: SUCCESS: no leaks!
last alloc =0x107f90
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
